### ISSUE TITLE
Load times have increased dramatically since upgrade to 1.4.x

### ISSUE THREAD

**Issue by bobvoorneveld** [cyberdelia/django-pipeline#428]
Hello,

We just tried upgrading django-pipeline from version 1.3.25 to 1.4.x (multiple minor versions tested, including latest: 1.4.6). Every request rendering static files went from 0.5 sec to about 20 secs to render. 

We tried to debug what led to this increase but could not find it. After downgrading to 1.3 the load times were again at 0.5 secs per page.

**Comment by snoepkast** [comment]
Can confirm, also found this: http://stackoverflow.com/questions/28239383/django-pipeline-page-load-really-slow

**Comment by mindcruzer** [comment]
+1

**Comment by ispivey** [comment]
+1 same problem. Rendering simple templates now consumes the majority of page load time, according to New Relic.

**Comment by mindcruzer** [comment]
@ispivey are you having problem in production?

**Comment by mindcruzer** [comment]
I believe the majority of this delay is happening because all static files are being copied to `STATIC_ROOT` every time a `{% stylesheet %}` or `{% javascript %}` tag is rendered, when in debug mode (ie. basically the equivalent of running `collectstatic` for every request). 

It would be much better to use the static files in their original locations when in debug mode, just like the default behavior with `runserver`. Of course, this presents a problem with compiling any assets, as the compiled versions will end up in the `STATICFILES_DIRS` folders.

**Comment by cyberdelia** [comment]
@mindcruzer is right.

This is debug/development mode, if you're running into this in production, you should fix your settings.

**Comment by mindcruzer** [comment]
Making the following [changes](https://github.com/mindcruzer/django-pipeline/commit/a401814d6bab729101d6506f0fd5c18aa8b697d9) decreases the load time dramatically in development (beware that I haven't had time to adequately test this). The only remaining delay seems to be with the compilers (lessc in my case). I am still getting ~1 sec page load times, which although a lot better than ~5 sec, is still enough to be annoying.

**Comment by ispivey** [comment]
I was able to address the issue in production -- I was using CachedStaticFilesStorage which resulted in a ton of network requests to memcached, and requests to S3 via S3BotoStorage when cache was cold, which added up. Moving to the newer ManifestStaticFilesStorage replaces all the round-trips with a local filesystem read.

My bad. Thanks for following up folks.

**Comment by cyberdelia** [comment]
@ispivey I will glady accept any PR improving our S3 documentation by introducting ``ManifestStaticFilesStorage`` in it :smile:

**Comment by ispivey** [comment]
@cyberdelia AM ON IT! :)

**Comment by ispivey** [comment]
@cyberdelia See https://github.com/cyberdelia/django-pipeline/pull/435

**Comment by snoepkast** [comment]
Since we use Heroku the `ManifestStaticFilesStorage` doesn't do the trick for us, and I don't see the fix from @mindcruzer in the official project. How do we proceed to fix this? Because for now it is too slow to be able to work with..

To clarify: yes this is in development with debug on. Since that was so slow we haven't tested it on a live environment yet.

**Comment by ispivey** [comment]
I am using ManifestStaticFilesStorage on Heroku. You're allowed to write to
the FS during collectstatic.

**Comment by snoepkast** [comment]
@ispivey I don't see why it should matter so much whether the manifest is in a file or in `redis` for example, and in fact it doesn't matter on my machine (still slow with `ManifestStaticFilesStorage`). It worked in previous versions of `django-pipeline`, and with nothing changed but the `django-pipeline` version (same `django` etc.) the behaviour changes, which indicates that it is indeed a problem of `django-pipeline` and not of the storage class in `django` itself.

If you want to push the `ManifestStaticFilesStorage` as the solution that would mean you have to run collectstatic on every deploy. Since if you don't, the manifest file doesn't end up in your slug, which will cause issues when you want to run multiple dyno's or do a restart. We don't want to do that because it means our deploy takes longer and we don't change static files every new release.

In short: I think `ManifestStaticFilesStorage` is not the solution for this problem, it doesn't fix it on my machine and it causes unwanted behaviour.

**Comment by Remiz** [comment]
I agree with snoepkast, django-pipeline is probably one of my favorite app that I install in most of my projects (I seriously think it would deserve to be integrated within Django, like RoR has the assets pipeline and manifest...), it makes things so much easier.

...But since the 1.4 I also noticed the performance issues in development (I have about 35 css files and 50 js/coffeescript files in my config dictionaries). It's true that it's behaving flawlessly on production, but the development mode need some love too. I recently started a new project without pipeline and saw the difference in loading time... It made me seriously reconsider using pipeline on it (or maybe an old version <1.4.*). It would be a shame to scare away new users from that fantastic project because it makes their dev environment slower...

Sorry to not propose a more concrete fix, I'm a bit too in-experienced to find a fix for the underlying cause, but I wanted to share my opinion on that issue.

**Comment by gone** [comment]
Yeah -  +1 on getting dev speedy. django-pipeline is a really great tool and I wouldn't move away from it  for (almost) anything, but dev is noticeably slower as is.

**Comment by toabi** [comment]
Jep, +1 for dev-speed. I have a project in a little Vagrant-VM, and on 1.4 it's not possible to develop with it anymore when with 1.3 there was no issue.

**Comment by Remiz** [comment]
@cyberdelia have you ever consider doing a kickstarter to fund your development time (and other contributors time too)? I know I would be happy to contribute financially to the project, and I'm sure I'm not the only one... Django Rest Framework did it and it was pretty successful (https://www.kickstarter.com/projects/tomchristie/django-rest-framework-3).

**Comment by gone** [comment]
I would kick money in for sure

**Comment by snoepkast** [comment]
I would too!

@cyberdelia: Are you working on a fix for this? If so do you have an ETA?
@mindcruzer: Did you manage to work a complete fix? If so can you create a pull request?

**Comment by cyberdelia** [comment]
I'm not actively working on this.

What is happening in dev mode is [fairly straightforward](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/templatetags/pipeline.py#L40-L46).

The [collector](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/collector.py) mimics collectstatic by copying _changed_ files to ``STATIC_ROOT``, and then the compilers are run on theses files.

So the slow down could come from theses two parts, if you can collect more info about what is effectively slow, that would help a lot.

**Comment by mindcruzer** [comment]
@cyberdelia There are two main issues, from what I can see:

1. [This](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/collector.py#L54-L76) is fairly slow. On my machine (brand new MBP), it takes about 0.0018 seconds to execute per file, from what I calculated. I counted 764 static files in my project, which takes the total time elapsed to about 1.4s, each time a pipeline template tag is rendered. 

2. The above is happening once per pipeline tag in the template. On the particular page I'm loading, there are three, and the total time to render the page is about 5.5s. 1.4s x 3 = 4.2s just to check if the static files need to be copied (in this example none of them do). The remaining 1.3s of latency I assume is coming from the compilers or other Django things. Indeed, if I set `def delete_file(...): return False`, my page load times drop to 1.3s - 1.4s, much closer to what I would reasonably expect. 

So it looks like checking if the file needs to be copied is what is making it slow, at least for me.

**Comment by mindcruzer** [comment]
Also, my guess is that this is mostly the fault of my vagrant VM, not pipeline (but I haven't yet verified this).

**Comment by mindcruzer** [comment]
Confirmed that enabling [NFS](http://docs.vagrantup.com/v2/synced-folders/nfs.html) in my Vagrant VM resolves the issue. Not an issue with pipeline after all.

**Comment by Remiz** [comment]
@mindcruzer I used to work on a Vagrant VM with NFS too and I switched to a local development setting on my mac for performance reasons. I have everything on a Mac OS X 10.7 / SSD machine with virtualenv installed (via brew), and it's still pretty slow... Page load when using a template that include a {% load pipeline %} takes between 1 to 4 seconds... When ajax calls take less than 200ms (because these do not use any assets). So as much as I agree that Vagrant without NFS makes everything super slow, I still think there is an issue with pipeline.

For a speed comparison, I recommend opening your browser developer tool and look at the network loading timeline. Check how long it takes to load a page using pipeline in dev mode, then compare to another page that doesn't use it (for instance the admin if you use django.contrib.admin), the difference for me is pretty noticeable.

**Comment by cyberdelia** [comment]
@mindcruzer Thanks a lot for taking the time to investigate, I really appreciate it. 

@Remiz Could you also conduct a deep dive into the same piece of code, and report what is actually slow for you? Thanks in advance.

**Comment by snoepkast** [comment]
I'm not working in a VM either, so I agree with @Remiz.

Will look more into it today (if my boss let's me).

**Comment by mindcruzer** [comment]
@Remiz It is still fairly slow compared to not using pipeline, I will admit. I am still getting ~500ms delay from checking static files to copy, but I am going to assume that this is largely latency from NFS. 

Aside from that, I am using the LESS compiler, and it is causing a further ~1.2s delay [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/compilers/less.py#L23), and more specifically [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/compilers/__init__.py#L98), which is just the `lessc

**Comment by mindcruzer** [comment]
@Remiz It is still fairly slow compared to not using pipeline, I will admit. I am still getting ~500ms delay from checking static files to copy, but I am going to assume that this is largely latency from NFS. 

Aside from that, I am using the LESS compiler, and it is causing a further ~1.2s delay [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/compilers/less.py#L23), and more specifically [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/compilers/__init__.py#L98), which is just the `lessc` command being run. What is more important to note is that its being run for every request. @cyberdelia is there a reason this is being run for every request, in contrast to some other compilers like [ES6](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/compilers/es6.py#L14-L15)? It seems to be on purpose.

**Comment by snoepkast** [comment]
Ok what I found:

Every template tag collects static all over again, which in my case is three times. This should only be done once per request. I had a quickfix but that would break if you have changes in your static files and not your python files (original code [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/collector.py#L30)):

```python
class Collector(object):
    found_files = OrderedDict()

    def collect(self):
        if not self.found_files:
            for finder in finders.get_finders():
                ...
```

If you write it as a middleware of some sorts that can be fixed. But I'm not sure if that's the way to go.

The other thing I found is that `django.contrib.staticfiles.finders.FileSystemFinder` is by far the slowest staticfiles finder on my machine (about 1 second outside of a VM). This is because it compares all files to the ignore patterns passed to the list functions, when those patterns are not given, the finder is almost instant, like the other finders. That would look something like this (original code [here](https://github.com/cyberdelia/django-pipeline/blob/master/pipeline/collector.py#L36)):

```python
for path, storage in finder.list([]):
```

**Comment by snoepkast** [comment]
Oh wait I just realised. Regarding the first 'fix'.

That might just work since all it does is return a collection of filenames and if you want to add a file or change a filename that would mean you have to change your bundles which would mean the dev server restarts and the 'cache' is cleared.

You probably know the logic better than I do, what do you think?

**Comment by mindcruzer** [comment]
@snoepkast Still poses a problem when editing existing static files

**Comment by Remiz** [comment]
Sorry for the delay, long weekend...
My findings are pretty similar to @snoepkast ... The collect is being called 4 times in my case, I guess that's for each inclusion of the template tag. I guess the fact that I'm using bower doesn't help to make it smaller either (it includes everything from the javascript libraries... test files, non-minified versions, docs...), because the number of files I see for the FileSystemFinder is 853 (and 820 for the AppDirectoriesFinder), and that being processed 4 times, that must be why it is slow...

I don't understand all the internals behind the collector mechanism, but I was thinking that during development the files will be read directly from their original location (within the app folders, or from the bower directory for instance...). Does it make sense to copy it each time in development? I remember in 1.3 that my .coffeescript files were generating a .js files at their original location.

**Comment by rizumu** [comment]
@Remiz with Bower, you should use the Pipeline finders to exclude unnecessary files, not the standard Django ones.

```
STATICFILES_FINDERS = [
    'pipeline.finders.FileSystemFinder',
    'pipeline.finders.AppDirectoriesFinder',
    'pipeline.finders.CachedFileFinder',
    'pipeline.finders.PipelineFinder',
]
```
http://django-pipeline.readthedocs.org/en/1.4.7/storages.html#using-with-staticfiles

**Comment by Remiz** [comment]
@rizumu Thanks for the tip, I'm ashamed to have missed that part in the doc... Now my numbers dropped, FileSystemFinder:266, AppDirectoriesFinder:465. It did impact the load time by about 400ms, so now my pages load in 1.5 seconds, which is still slower than without without pipeline, but slightly better than before.

**Comment by snoepkast** [comment]
@cyberdelia what do you think about https://github.com/wearespindle/django-pipeline/commit/934708870db8b93e99800e084a7c722bf5bdbe1c? It's a very basic fix for preventing the execution of `collect` multiple times per request.

You'll have to put `django.core.context_processors.request` in your `TEMPLATE_CONTEXT_PROCESSORS`, but it still works without that processor, then it simply collects for every template tag call.

**Comment by pdr** [comment]
+1 also having performance issue. Also sad because I think pipeline is great but I can't continue using it like this. 

The issue is that it now runs default_collector.collect(), even when DEBUG = True

**Comment by jaddison** [comment]
I recently started using `pipeline` in a few of my Django projects, I like how it works, but there certainly is a marked slowdown in pages loaded during development. Is this something that is actually fixable, or has `pipeline` moved in a direction where this won't be fixed?

**Comment by cyberdelia** [comment]
@jaddison Not really, are you using SASS too ?

**Comment by jaddison** [comment]
No SASS, just LESS.

**Comment by pdr** [comment]
I resolved my issues by reducing the amount of files that were included in my static directories. Load times still slower than before, but more reasonable.

**Comment by snoepkast** [comment]
I would appreciate it if @cyberdelia would take a look at my change and gave his opinion on whether I'm going in the right direction with it.

**Comment by skolsuper** [comment]
Rendering a simple login page with a single css file now takes 20s in my dev environment, compared to <200ms on `1.3.27`. I can't even test the rest of my site using the devserver.

@cyberdelia I am using SASS, your question above implies this might be a factor, is there already a known fix I don't know about?

**Comment by leonardoo** [comment]
if we remove default_collector.collect(), from PipelineMixin, pipeline keep working but load time improved significantly

**Comment by nav** [comment]
Switching to `pipeline.finders.FileSystemFinder` and `pipeline.finders.AppDirectoriesFinder` STATICFILES_FINDERS helped me.

**Comment by Remiz** [comment]
It seems fixed to me thanks to #455 (nice job!). Does anyone else confirm?
